Fix Incorrect Payout model field, destination

arnav13081994 opened the pull request:
<!-- Thank you for helping us out: your contribution means a great deal to the project and the community as a whole! -->


## Description

<!-- What are you proposing? -->

This PR contains the following changes:

1. `destination` on the `Payout` model can **either** refer to the `Card` or the `BankAccount` object but was incorrectly referenced as a `FK` to the `BankAccount` object.
2. Added `Data Migration` to correctly migrate data from `BankAccount.djstripe_id (pk)` to `DjstripePaymentMethod.id (pk)` and vice-versa. 
3. Ensured `migrations` are completely `reversible` as well. And split `migrations` over 3 files to avoid the `Pending Trigger Events Error`.
4. Added Corresponding Tests.


Checklist:

- [X] I've updated the `tests` or confirm that my change doesn't require any updates.
- [X] I've updated the `documentation` or confirm that my change doesn't require any updates.
- [X] I confirm that my change doesn't drop code coverage below the current level.
- [X] I've updated `migrations` or confirm that my change doesn't make changes to any model.

## Rationale

<!-- 
Why does this project need the change you're proposing? 
If this pull request fixes an open issue, don't forget to link it with `Fix #NNNN` 
-->

`Payout` objects with `destination=card` will now get synced correctly.

Fix: #1826

arnav13081994 committed:
Added Payout Fixtures

arnav13081994 committed:
Fixed incorrect destination model field of Payout model

destination can either be a Card or a Bank Account object but was incorrectly being
references to the BankAccount object.

aburgel commented on djstripe/migrations/0019_alter_payout_destination_clone.py:
```diff
@@ -0,0 +1,43 @@
+            LOOP
+                UPDATE djstripe_payout SET destination_clone_id = (SELECT id FROM djstripe_bankaccount WHERE djstripe_id::VARCHAR = destination_pk);
+            END LOOP;
+            END$$;
```
```suggestion
UPDATE djstripe_payout SET destination_clone_id = (select id from djstripe_bankaccount where djstripe_bankaccount.djstripe_id = djstripe_payout.destination_id)
WHERE EXISTS(select * from djstripe_bankaccount where djstripe_bankaccount.djstripe_id = djstripe_payout.destination_id);
```

I'm not sure if that SQL will run on all engines. I did a bit of research to find the ANSI SQL syntax for an UPDATE that effectively joins another table. This is what I came up with.

I've tested on postgres and sqlite.

aburgel commented on djstripe/migrations/0019_alter_payout_destination_clone.py:
```diff
@@ -0,0 +1,43 @@
+                RAISE NOTICE '%', (SELECT djstripe_id FROM djstripe_bankaccount WHERE id::VARCHAR = destination_clone_pk);
+                UPDATE djstripe_payout SET destination_id = (SELECT djstripe_id FROM djstripe_bankaccount WHERE id::VARCHAR = destination_clone_pk);
+            END LOOP;
+            END$$;
```
```suggestion
UPDATE djstripe_payout SET destination_id = (select djstripe_id from djstripe_bankaccount where djstripe_bankaccount.id = djstripe_payout. destination_clone_id)
WHERE EXISTS(select * from djstripe_bankaccount where djstripe_bankaccount.id = djstripe_payout.destination_clone_id);
```

aburgel reviewed (commented):
To safely migrate this data, we'll need some additional work. At the very least, you will need to ensure that nothing updates the table until the migrations are complete and the app is updated.

If updates come in after the data is copied over in 19, but before the running code is updated, then the old destination_id will be written to, and then will never be copied over to the new column.

Similarly, running 20 will remove a column that is in active use (until the new code is deployed), and that could lead to errors. Though errors are better than bad data.

Some other options:
1. Add the new column and provide a command for the user to run after the code is deployed that copies over the data. Then update the code to read from both columns so it handles the case before and after the data is copied. Finally, in a subsequent release we can clean it all up.
2. We can ask people to deploy during a maintenance window when their app is offline.
3. Provide some way to sync the payout table from stripe after the migration is complete, so any dropped updates get corrected.

arnav13081994 committed:
Added CUSTOM_SQL Dict to track forward and backward sql for supported dbs

CUSTOM_SQL will be used to keep track of the custom SQL required to be run
for each of the 3 dj-tripe supported databases.

arnav13081994 committed:
Added a helper method get_sql_for_connection

get_sql_for_connection is used to pick the correct
collection of SQL statements that need to run for the
given SQL engine and migration direction.

arnav13081994 committed:
Refactored SQL for every DB engine and migration direction into separate files

    This was done in order to separately manage and maintain the SQL
    and to eventually consume the SQL from files instead of
    directly from a multi line string.

arnav13081994 committed:
Updated get_sql_for_connection to read from SQL files

jleclanche commented on djstripe/migrations/0013_alter_payout_destination.py:
```diff
@@ -0,0 +1,33 @@
+    vendor, sql_statement = get_sql_for_connection(schema_editor, "backward")
+    with schema_editor.connection.cursor() as cursor:
+        if vendor == "sqlite":
+            cursor.executescript(sql_statement)
```
Use `schema_editor.execute`. cf https://stackoverflow.com/questions/32384547/django-migration-runsql-conditional-on-database-type

arnav13081994 replied:
That is not [recommended](https://docs.djangoproject.com/en/4.1/ref/migration-operations/#runpython):

> By default, RunPython will run its contents inside a transaction on databases that do not support DDL transactions (for example, MySQL and Oracle). This should be safe, but may cause a crash if you attempt to use the schema_editor provided on these backends; in this case, pass atomic=False to the RunPython operation.

Would need to tweak the `atomic` kwarg as well. 

Django docs themselves [recommend](https://docs.djangoproject.com/en/4.1/topics/db/sql/#executing-custom-sql-directly) using `cursor` to execute custom SQL queries directly.

jleclanche commented on sql/utils.py:
```diff
@@ -0,0 +1,25 @@
+from pathlib import Path
+
+
+def get_sql_for_connection(schema_editor, direction: str) -> str:
```
I appreciate what you're trying to do here but let's move it to the migrations folder. This sql module should not exist.

Just move this function to the migration itself, and have the sql files in `djstripe/migrations/sql/0013_postgres_forwards.sql` etc.

arnav13081994 replied:
@jleclanche Changes made. Please re-review.

arnav13081994 committed:
Moved get_sql_for_connection from utils to migration 0013_alter_payout_destination

arnav13081994 committed:
Moved all custom migrations to djstripe/migrations/sql folder
